### PR TITLE
DEPR: Move Numericindex._convert_slice_indexer & ._maybe_cast_slice_bound

### PR DESCRIPTION
**Pull request by topper-123**
Moves`_convert_slice_indexer` & `_maybe_cast_slice_bound` from `NumericIndex`to `Index` in preparation to remove `NumericIndex` and include numpy int/uint/float64 in the base `Index`.

Each commit contains one method moving each, could be easier to view each method separately in the commit. 

xref #42717.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3983,6 +3983,13 @@ def _convert_slice_indexer(self, key: slice, kind: str_t):
         # potentially cast the bounds to integers
         start, stop, step = key.start, key.stop, key.step
 
+        # TODO(GH#50617): once Series.__[gs]etitem__ is removed we should be able
+        #  to simplify this.
+        if is_float_dtype(self.dtype):
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:3988]: need to specify numpy dtype

**Comment by topper-123** [comment]
Updated.

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx @topper-123

**Pull request merged**